tool/tsh/common/db_exec.go

```diff
@@ -0,0 +1,647 @@
+	"golang.org/x/sync/errgroup"
+
+	"github.com/gravitational/teleport"
+	apiclient "github.com/gravitational/teleport/api/client"
+	"github.com/gravitational/teleport/api/client/proto"
+	apidefaults "github.com/gravitational/teleport/api/defaults"
+	"github.com/gravitational/teleport/api/types"
+	apiutils "github.com/gravitational/teleport/api/utils"
+	"github.com/gravitational/teleport/lib/client"
+	"github.com/gravitational/teleport/lib/client/db/dbcmd"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/srv/alpnproxy"
+	"github.com/gravitational/teleport/lib/tlsca"
+	"github.com/gravitational/teleport/lib/utils"
+	logutils "github.com/gravitational/teleport/lib/utils/log"
+	"github.com/gravitational/teleport/tool/common"
+)
+
+func onDatabaseExec(cf *CLIConf) error {
+	execCommand, err := newDatabaseExecCommand(cf)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer execCommand.close()
+
+	return trace.Wrap(execCommand.run())
+}
+
+// databaseExecClient is a wrapper of client.TeleportClient that makes backend
+// calls. can be mocked for testing.
+type databaseExecClient interface {
+	close() error
+	getProfileStatus() *client.ProfileStatus
+	getAccessChecker() services.AccessChecker
+	issueCert(context.Context, *databaseInfo) (tls.Certificate, error)
+	listDatabasesWithFilter(context.Context, *proto.ListResourcesRequest) ([]types.Database, error)
+}
+
+type databaseExecCommand struct {
+	cf          *CLIConf
+	tc          *client.TeleportClient
+	client      databaseExecClient
+	makeCommand func(context.Context, *databaseInfo, string, string) (*exec.Cmd, error)
+	summary     databaseExecSummary
+}
+
+func newDatabaseExecCommand(cf *CLIConf) (*databaseExecCommand, error) {
+	if err := checkDatabaseExecInputFlags(cf); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	tc, err := makeClient(cf)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	sharedClient, err := newSharedDatabaseExecClient(cf, tc)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	commandMaker, err := newDatabaseExecCommandMaker(cf.Context, tc, sharedClient.getProfileStatus())
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &databaseExecCommand{
+		cf:          cf,
+		tc:          tc,
+		client:      sharedClient,
+		makeCommand: commandMaker.makeCommand,
+	}, nil
+}
+
+func (c *databaseExecCommand) run() error {
+	// Fetch.
+	dbs, err := c.getDatabases()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Execute  in parallel.
+	group, groupCtx := errgroup.WithContext(c.cf.Context)
+	group.SetLimit(c.cf.ParallelJobs)
+	for _, db := range dbs {
+		group.Go(func() error {
+			result := c.exec(groupCtx, db)
+			c.summary.add(result)
+			return nil
+		})
+	}
+
+	// Print summary.
+	defer func() {
+		switch {
+		case c.cf.OutputDir != "":
+			c.summary.printAndSave(c.cf.Stdout(), c.cf.OutputDir)
+		case len(dbs) > 1:
+			c.summary.print(c.cf.Stdout())
+		}
+	}()
+
+	return trace.Wrap(group.Wait())
```

Comment:
Is there any sort of timeout on the execution?

Reply:
no. the timeout will vary depending on the command to be executed.  we can add a new flag in the future if a user asks for it. for now, user can always ctrl-c to kill it.